In [72]:
#1
import numpy as np
def my_bin_2_dec(b):
    d = 0
    length = len(b)
    i = 0
    while i < length:
        if b[i] == 1:
            d += pow(2, length - 1 - i)
        i += 1
        #print(d)
    return d

print(my_bin_2_dec([1,1,1,1,1,1,1,1,1,1,0]))

2046


In [74]:
#2
def my_dec_2_bin(d):
    if d == 0:
        return [0]
    length = int(np.log2(d) + 1) # Found this by recognizing that d <= 2^(n_max + 1) = 2^length
    b = np.full(length, 0, dtype = int)
    i = 0
    while i < length:
        if d / pow(2, length - 1 - i) >= 1:
            b[i] = 1
            d -= pow(2, length - 1 - i)
        i += 1
    return b

print(my_dec_2_bin(16000))

[1 1 1 1 1 0 1 0 0 0 0 0 0 0]


In [45]:
#3
print(my_bin_2_dec(my_dec_2_bin(12654)))
#Yes I do get the same number

12654


In [62]:
#4
def my_bin_adder(b1, b2):
    length1 = len(b1)
    length2 = len(b2)
    if length1 - length2 > 0: #Make the output array length 1 larger than the largest input array
        length = length1 + 1  
    else:
        length = length2 + 1

    x = length1 #Make both input arrays the same length
    y = length2
    while x < length:
        b1 = [0] + b1
        x += 1
    while y < length:
        b2 = [0] + b2
        y += 1
    
    b = np.full(length, 0, dtype = int)
    i = 0
    while i < length: #Do some binary adding
        if b1[length - 1 - i] + b2[length - 1 - i] + b[length - 1 - i] == 1:
            b[length - 1 - i] = 1
        elif b1[length - 1 - i] + b2[length - 1 - i] + b[length - 1 - i] == 2:
            b[length - 1 - i] = 0
            b[length - 2 - i] = 1
        elif b1[length - 1 - i] + b2[length - 1 - i] + b[length - 1 - i] == 3:
            b[length - 1 - i] = 1
            b[length - 2 - i] = 1
        i += 1

    if b[0] == 0:
        b = np.delete(b, 0)
    return b

print(my_bin_adder([1, 1, 1, 1, 1], [1, 0, 1, 0, 1, 0, 0]))

[1 1 1 0 0 1 1]


#5
Adding more bits to the sign is obviously just useless / worse than useless, since it is a binary operation. Adding bits to 
the characteristic allows for larger numbers to be producible, and adding bits to the fraction makes the spacing between 
values smaller (relativeto the size of the number that is being observed).

In [30]:
#6
def my_ieee_2_dec(ieee):
    arr = [int(x) for x in ieee] #Make into array of 0s and 1s
    s = arr[0]
    
    e = 0
    i = 1
    while i < 12:
        if arr[i] == 1:
            e += pow(2, 11 - i)
        i += 1
    
    f = 0
    while i < 64:
        if arr[i] == 1:
            f += 1 / pow(2, i - 11)
        i += 1

    d = pow(-1, s) * pow(2, e - 1023) * (1 + f)        
    return d

d = my_ieee_2_dec('0100000000101110010111101010001110011100001100011010010001101000')
#print("{:.29f}".format(d)) #Necessary if you want to see down to a certain number of sig-figs
print(d)

15.18484199625


In [39]:
#7
import numpy as np
def my_dec_2_ieee(d):
    ieee = [0] * 64
    if d < 0:
        ieee[0] = 1
    elif d > 0:
        ieee[0] = 0

    i = 1
    e = 0
    exp = np.log2(np.abs(d))
    while i < 12:
        if exp > pow(2, 11 - i) + e - 1023:
            e += pow(2, 11 - i)
            ieee[i] = 1
        i += 1

    d2 = pow(2, e - 1023)
    f = 0
    while i < 64:
        if d2 * (1 + f + pow(2, -(i - 11))) < np.abs(d):
            f += pow(2, -(i - 11))
            ieee[i] = 1
        elif d2 * (1 + f + pow(2, -(i - 11))) == np.abs(d):
            f += pow(2, -(i - 11))
            ieee[i] = 1
            break
        i += 1
    
    return ieee

ieee = my_dec_2_ieee(-309.141740)
string = ''
for i in ieee:
    string += str(i)
print(string)

1100000001110011010100100100010010010001001010011000100010010000


In [42]:
#8
%reset -f
import numpy as np
def ieee_baby(d):
    ieee = [0] * 6
    if d < 0:
        ieee[0] = 1
    elif d > 0:
        ieee[0] = 0

    i = 1
    e = 0
    exp = np.log2(np.abs(d))
    while i < 3:
        if exp > pow(2, 2 - i) + e - 1:
            e += pow(2, 2 - i)
            ieee[i] = 1
        i += 1

    d2 = pow(2, e - 1)
    f = 0
    while i < 6:
        if d2 * (1 + f + pow(2, -(i - 2))) < np.abs(d):
            f += pow(2, -(i - 11))
            ieee[i] = 1
        elif d2 * (1 + f + pow(2, -(i - 2))) == np.abs(d):
            f += pow(2, -(i - 2))
            ieee[i] = 1
            break
        i += 1
    
    return ieee

ieee = ieee_baby(1)
string = ''
for i in ieee:
    string += str(i)
print(string)

000100


Tried messing around with the code to make it work for 6 bit, but I couldn't figure it out.
However, in general there are 3 possible values of the characteristic part: 1, 2, and 3.
There are 7 possible values for the fraction, which are 0.5, 0.25, 0.125, 0.75, 0.625, 0.375, and 0.875.
This allows for the following possibilities:

In [50]:
arr1 = [1, 2, 3]
arr2 = [0.5, 0.25, 0.125, 0.75, 0.625, 0.375, 0.875]

i = 0
while i < 3:
    j = 0
    while j < 7:
        a = arr1[i] * arr2[j]
        print(a)
        j += 1
    i += 1
        

0.5
0.25
0.125
0.75
0.625
0.375
0.875
1.0
0.5
0.25
1.5
1.25
0.75
1.75
1.5
0.75
0.375
2.25
1.875
1.125
2.625


This leaves (in ascending order and after removing duplicates): 
0.125, 0.25, 0.375, 0.5, 0.625, 0.75, 0.875, 1.0, 1.125, 1.25, 1.5, 1.75, 1.875, 2.25, and 2.625
Which means the smallest gap is 0.125 and the largest is 0.375

In [70]:
#9
%reset -f
import numpy as np
i = 4503599627300000
while i > -1:
    if np.spacing(i) >= 1:
        print(i)
        break
    i += 1

4503599627370496 1.0


By altering the incrementing of i, I was able to find this. I started i=1 and multiplied by 2 with every iteration,
then I started from 4E15 and did i += 1000000, then I started at 4.4...E15 and did i += 10000, then i += 1 from 
4503599627300000. The final answer is 4503599627370496.

#10
Well binary is more efficient computationally, and logically I suppose it may be marginally more efficient for
everyday use as well, though to a much less significant degree (we aren't doing millions of calculations a second).
However, numbers take less physical space to write in base10, for example, 16 in base10 is 10000 in binary.
Base10 also allows for greater distinction in orders of magnitude, though I don't really know if this is necessarily
a good thing.

#11
0000000000000 would be 13 in base1, since it is literally just the number in base10 written as individual
0s for every integer in [1,13]. To add, you just add the number of 0s in both numbers and smush them together.
To multiply, for every 0 in the first number write however many zeros there are in the second until you've gone
through all the zeros in both.

#12
Well I've got ten fingers, so that's 10 1s, or 1111111111, which according to my binary number calculator is 1023

#13
Just like how in base10 you can multiply by 10 by just adding a zero at the end of the number (or technically shifting
the decimal place once to the right), the same thing can happen when multiplying by 2 in base2, i.e. 2 : 1, 4 : 10, 8 : 100, etc.
I made sure this was the case with my binary number calculator and it is so.